hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/listener/ClientSmartListenerService.java

```diff
@@ -19,102 +19,116 @@
-import com.hazelcast.util.ConstructorFunction;
-import com.hazelcast.util.EmptyStatement;
 import com.hazelcast.util.UuidUtil;
 
 import java.util.Collection;
 import java.util.Collections;
-import java.util.Iterator;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.Future;
 
-public class ClientSmartListenerService extends ClientListenerServiceImpl {
+public class ClientSmartListenerService extends ClientListenerServiceImpl implements InitialMembershipListener {
 
-    private static final ConstructorFunction<ClientRegistrationKey, Set<ClientEventRegistration>> CONSTRUCTOR_FUNCTION
-            = new ConstructorFunction<ClientRegistrationKey, Set<ClientEventRegistration>>() {
-        @Override
-        public Set<ClientEventRegistration> createNew(ClientRegistrationKey arg) {
-            return Collections.newSetFromMap(new ConcurrentHashMap<ClientEventRegistration, Boolean>());
-        }
-    };
-    private final ConcurrentMap<ClientRegistrationKey, Set<ClientEventRegistration>> registrations =
-            new ConcurrentHashMap<ClientRegistrationKey, Set<ClientEventRegistration>>();
-    private final Object regMutex = new Object();
+    private final Set<Member> members = new HashSet<Member>();
+    private final Map<ClientRegistrationKey, Map<Address, ClientEventRegistration>> registrations
+            = new ConcurrentHashMap<ClientRegistrationKey, Map<Address, ClientEventRegistration>>();
+    private final Object listenerRegLock = new Object();
 
-    public ClientSmartListenerService(HazelcastClientInstanceImpl client, int eventThreadCount, int eventQueueCapacity) {
+    public ClientSmartListenerService(HazelcastClientInstanceImpl client,
+                                      int eventThreadCount, int eventQueueCapacity) {
         super(client, eventThreadCount, eventQueueCapacity);
+        client.getClientClusterService().addMembershipListener(this);
     }
 
     @Override
-    public String registerListener(BaseClientAddListenerRequest request, EventHandler handler) {
-        request.setLocalOnly();
+    public String registerListener(BaseClientAddListenerRequest addRequest,
+                                   BaseClientRemoveListenerRequest removeRequest,
+                                   EventHandler handler) {
+        addRequest.setLocalOnly();
+        String userRegistrationId = UuidUtil.newUnsecureUuidString();
+        synchronized (listenerRegLock) {
 
-        Collection<Member> members = client.getClientClusterService().getMemberList();
+            ClientRegistrationKey registrationKey =
+                    new ClientRegistrationKey(userRegistrationId, addRequest, removeRequest, handler);
+            registrations.put(registrationKey, new ConcurrentHashMap<Address, ClientEventRegistration>());
+            try {
```

Comment:
it seems like we can release the lock at this point, before time consuming invoke calls.